File: compiler/ast_nodes.py
class Program:
    def __init__(self, statements): self.statements = statements
class VarDecl:
    def __init__(self, name, expr): self.name, self.expr = name, expr
class AssignStmt:
    def __init__(self, name, expr): self.name, self.expr = name, expr
class PrintStmt:
    def __init__(self, expr): self.expr = expr
class IfStmt:
    def __init__(self, cond, then_branch, else_branch=None):
        self.cond, self.then_branch, self.else_branch = cond, then_branch, else_branch
class ForStmt:
    def __init__(self, init, cond, update, body): self.init, self.cond, self.update, self.body = init, cond, update, body
class BinOp:
    def __init__(self, left, op, right): self.left, self.op, self.right = left, op, right
class Num:
    def __init__(self, value): self.value = value
class Var:
    def __init__(self, name): self.name = name

File: compiler/ast_nodes.ts
export abstract class ASTNode {}

export class Program extends ASTNode {
  constructor(public statements: Statement[]) {
    super();
  }
}

export abstract class Statement extends ASTNode {}

export class VarDecl extends Statement {
  constructor(public name: string, public value: Expression) {
    super();
  }
}

export class AssignStmt extends Statement {
  constructor(public name: string, public value: Expression) {
    super();
  }
}

export class PrintStmt extends Statement {
  constructor(public value: Expression) {
    super();
  }
}

export class IfStmt extends Statement {
  constructor(
    public condition: Expression,
    public thenBlock: Statement[],
    public elseBlock: Statement[] | null = null
  ) {
    super();
  }
}

export class ForStmt extends Statement {
  constructor(
    public init: Statement,
    public condition: Expression,
    public update: Statement,
    public body: Statement[]
  ) {
    super();
  }
}

export abstract class Expression extends ASTNode {}

export class BinOp extends Expression {
  constructor(public left: Expression, public op: string, public right: Expression) {
    super();
  }
}

export class UnaryOp extends Expression {
  constructor(public op: string, public operand: Expression) {
    super();
  }
}

export class Num extends Expression {
  constructor(public value: number) {
    super();
  }
}

export class Var extends Expression {
  constructor(public name: string) {
    super();
  }
}
